File: cover/src/main/java/com/ruidoespontaneo/cassette/cover/components/PreviewWaveform.kt
package com.ruidoespontaneo.cassette.cover.components

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.material3.ExperimentalMaterial3ExpressiveApi
import androidx.compose.material3.LinearWavyProgressIndicator
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.key
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

/** One line of the waveform: each gets its own wave shape and speed so the set reads as one waveform. */
private data class WaveLine(val wavelength: Dp, val waveSpeed: Dp, val amplitude: Float)

private val waveLines = listOf(
    WaveLine(wavelength = 40.dp, waveSpeed = 40.dp, amplitude = 1f),
    WaveLine(wavelength = 28.dp, waveSpeed = 64.dp, amplitude = 0.7f),
    WaveLine(wavelength = 52.dp, waveSpeed = 26.dp, amplitude = 0.85f)
)

/** How many lines [PreviewWaveform] draws, and so how many colors it can use. */
const val PREVIEW_WAVEFORM_LINES = 3

private const val IDLE_ALPHA = 0.35f

/** How much taller than its stroke a line's container is, leaving room for the wave's swing. */
private const val WAVE_CONTAINER_TO_STROKE = 2.75f

/**
 * A stack of horizontal Material 3 Expressive wavy lines, one per entry in [colors] (up to
 * [PREVIEW_WAVEFORM_LINES]; extra colors are ignored), each as wide as this composable is. While
 * [playing] the waves ripple; otherwise they ease down to flat, dimmed lines and the composable
 * keeps its size, so nothing around it shifts. [strokeWidth] is how thick each line is (Material's
 * default is 4.dp) and each line's container is sized from it so the waves have room; [lineSpacing]
 * is the gap between lines.
 *
 * Draws no background of its own — the colors are meant for whatever it sits on (see
 * [waveformColors], which takes that background).
 *
 * Decorative: it isn't driven by the audio itself — callers just say whether something is playing.
 */
@OptIn(ExperimentalMaterial3ExpressiveApi::class)
@Composable
fun PreviewWaveform(
    playing: Boolean,
    colors: List<Color>,
    modifier: Modifier = Modifier,
    strokeWidth: Dp = 8.dp,
    lineSpacing: Dp = 4.dp
) {
    val stroke = with(LocalDensity.current) { Stroke(width = strokeWidth.toPx(), cap = StrokeCap.Round) }
    val alpha by animateFloatAsState(targetValue = if (playing) 1f else IDLE_ALPHA, label = "waveformAlpha")
    Column(
        modifier = modifier.alpha(alpha),
        verticalArrangement = Arrangement.spacedBy(lineSpacing),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        colors.zip(waveLines).forEach { (color, line) ->
            // Keyed on [playing] so each state gets a fresh indicator that starts at the right
            // amplitude. The indicator eases amplitude changes itself but only starts a new easing
            // when none is running, so a quick on/off/on (switching tracks buffers in between) would
            // otherwise drop the last change and leave it flat while playing. The alpha fade above
            // covers for the amplitude snapping.
            key(playing) {
                LinearWavyProgressIndicator(
                    progress = { 1f },
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(strokeWidth * WAVE_CONTAINER_TO_STROKE),
                    color = color,
                    trackColor = Color.Transparent,
                    stroke = stroke,
                    gapSize = 0.dp,
                    stopSize = 0.dp,
                    amplitude = { if (playing) line.amplitude else 0f },
                    wavelength = line.wavelength,
                    waveSpeed = line.waveSpeed
                )
            }
        }
    }
}

File: iosApp/iosApp/Common/PreviewWaveform.swift
import SwiftUI

/// How many lines `PreviewWaveform` draws, and so how many colours it can use.
let previewWaveformLines = 3

/// A stack of horizontal wavy lines, one per entry in `colors` (up to `previewWaveformLines`), each
/// as wide as the view. While `playing` the waves ripple; otherwise they ease down to flat, dimmed
/// lines, and the view keeps its size so nothing around it shifts. A port of Android's
/// PreviewWaveform, which draws Material's wavy progress indicator with the same parameters.
///
/// Draws no background of its own: the colours are meant for whatever it sits on (see
/// `waveformColors`, which takes that background).
///
/// Decorative: it isn't driven by the audio itself. Callers just say whether something is playing.
struct PreviewWaveform: View {
    let playing: Bool
    let colors: [Color]
    var strokeWidth: CGFloat = 8
    var lineSpacing: CGFloat = 4

    /// One line of the waveform: each gets its own wave shape and speed, so the set reads as one
    /// waveform.
    private struct Line {
        let wavelength: CGFloat
        let speed: CGFloat
        let amplitude: CGFloat
    }

    private static let lines = [
        Line(wavelength: 40, speed: 40, amplitude: 1),
        Line(wavelength: 28, speed: 64, amplitude: 0.7),
        Line(wavelength: 52, speed: 26, amplitude: 0.85),
    ]
    private static let idleAlpha = 0.35
    /// How much taller than its stroke a line's frame is, leaving room for the wave's swing.
    private static let frameToStroke: CGFloat = 2.75

    var body: some View {
        WaveAnimation(running: playing) { amplitude, time in
            VStack(spacing: lineSpacing) {
                ForEach(Array(zip(colors, Self.lines).enumerated()), id: \.offset) { _, pair in
                    let (color, line) = pair
                    WaveLine(
                        amplitude: amplitude * line.amplitude,
                        phase: wavePhase(time: time, speed: line.speed, wavelength: line.wavelength),
                        wavelength: line.wavelength,
                        strokeWidth: strokeWidth
                    )
                    .stroke(color, style: StrokeStyle(lineWidth: strokeWidth, lineCap: .round))
                    .frame(height: strokeWidth * Self.frameToStroke)
                }
            }
        }
        .opacity(playing ? 1 : Self.idleAlpha)
        .animation(.easeInOut(duration: 0.3), value: playing)
        .accessibilityHidden(true)
    }
}

/// One sine line across the frame, centred vertically. `amplitude` is a 0...1 fraction of the room
/// the frame leaves beside the stroke, and animates.
private struct WaveLine: Shape {
    var amplitude: CGFloat
    let phase: CGFloat
    let wavelength: CGFloat
    let strokeWidth: CGFloat

    var animatableData: CGFloat {
        get { amplitude }
        set { amplitude = newValue }
    }

    func path(in rect: CGRect) -> Path {
        // Round caps reach half a stroke past each end, so the line starts and stops that far in.
        let start = rect.minX + strokeWidth / 2
        let end = rect.maxX - strokeWidth / 2
        let swing = max(rect.height - strokeWidth, 0) / 2 * amplitude
        var path = Path()
        var x = start
        path.move(to: point(x, rect: rect, swing: swing))
        while x < end {
            x = min(x + 2, end)
            path.addLine(to: point(x, rect: rect, swing: swing))
        }
        return path
    }

    private func point(_ x: CGFloat, rect: CGRect, swing: CGFloat) -> CGPoint {
        CGPoint(x: x, y: rect.midY + swing * sin(2 * .pi * (x - phase) / wavelength))
    }
}
